File: app/models/community_wire.rb
class CommunityWire

  extend ActsAsApi::Base

  def model_name
    "community_wire"
  end
  
  def initialize(community)
    @community = community
  end

  def events
    @community.events.upcoming.reorder("date ASC").limit(3)
  end

  def transactions
    @community.transactions.reorder("created_at DESC").limit(3)
  end

  def neighborhood
    @community.posts.where(category: 'neighborhood').limit(3)
  end

  def offers
    @community.posts.where(category: 'offers').limit(3)
  end

  def help
    @community.posts.where(category: 'help').limit(3)
  end

  def publicity
    list = (@community.posts.where(category: 'publicity').limit(3) +
      @community.announcements.reorder("GREATEST(replied_at, created_at) DESC").limit(3))
      .sort {|a,b| [b.replied_at, b.created_at].compact.max <=> [a.replied_at, a.created_at].compact.max }.take(3) 
  end
  
  def meetups
    @community.posts.where(category: "meetups").limit(3)
  end

  def group
    GroupPost.order("group_posts.updated_at DESC").includes(:group).
      where(groups: { community_id: @community.id }).limit(3)
  end

  def other
    @community.posts.where(category: 'other').limit(3)
  end

  acts_as_api
  
  api_accessible :default do |t|
    t.add :events
    t.add :neighborhood
    t.add :offers
    t.add :help
    t.add :publicity
    t.add :group
    t.add :announcements
    t.add :other
    t.add :past
    t.add :meetups
    t.add :transactions
  end
  
end

File: app/javascripts/models/community_wire.js
var CommunityWire = Model.extend({
  initialize: function(options) {
    this.uri = "/api" + options.uri;
  },
  
  fetch: function(data, callback) {
    var self = this;
    $.getJSON(this.uri, data, function(response) {
      self.neighborhood = new Posts(response["neighborhood"], CommonPlace.community.categories.neighborhood);
      self.offers = new Posts(response["offers"], CommonPlace.community.categories.offers);
      self.help = new Posts(response["help"], CommonPlace.community.categories.help);
      self.publicity = new PostLikes(response["publicity"], CommonPlace.community.categories.publicity);
      self.other = new Posts(response["other"], CommonPlace.community.other);

      self.groupPosts = new GroupPosts(response["group"], CommonPlace.community.groupPosts);
      self.events = new Events(response["events"], CommonPlace.community.events);
      self.meetups = new Posts(response["meetups"], CommonPlace.community.categories.meetups);
      self.transactions = new Transactions(response["transactions"], CommonPlace.community.transactions);
      if (callback) { callback() };
    });
  },
  
  all: function() { return [
    this.neighborhood, this.offers, this.help, this.publicity, this.other, this.groupPosts, this.events, this.meetups, this.transactions];
  }
});

